pad_server/src/state.c

```diff
@@ -40,15 +43,85 @@ int padstate_change_level(padstate_t *state, arm_lvl_e new_arm) {
 
     return pthread_rwlock_unlock(&state->rw_lock);
 }
-/* TODO: docs
- * NOTE: This does not check for valid actuator ID
+
+/*
+ * Set the value of an actuator with the required progression
+ * @param id The actuator id
+ * @param req_state The new actuator state
+ * @return ACT_OK for success, ACT_DNE for invalid id, ACT_INV for invalid req_state, -1 for errors eith errno being
+ * set
  */
-int padstate_actuate(padstate_t *state, uint8_t id, bool new_state) {
+int padstate_actuate(padstate_t *state, uint8_t id, uint8_t req_state) {
+    if (id >= NUM_ACTUATORS) return ACT_DNE;
+
+    if (req_state != 0 && req_state != 1) return ACT_INV;
+    bool new_state = (bool)req_state;
+
+    arm_lvl_e arm_lvl;
+    int err = padstate_get_level(state, &arm_lvl);
+    if (err) {
+        errno = err;
+        return -1;
+    }
+
+    bool is_solenoid_valve = id >= ID_XV1 && id <= ID_XV12;
+    fprintf(stderr, "%ul\n", arm_lvl);
+
+    switch (arm_lvl) {
+    case ARMED_PAD:
+        return ACT_DENIED;
+        break;
+
+    case ARMED_VALVES:
+        if (!is_solenoid_valve) {
+            return ACT_DENIED;
+        }
+        break;
+
+    case ARMED_IGNITION:
+        if (!is_solenoid_valve && id != ID_QUICK_DISCONNECT) {
+            return ACT_DENIED;
+        }
+        break;
+
+    case ARMED_DISCONNECTED:
+        if (!is_solenoid_valve && id != ID_QUICK_DISCONNECT && id != ID_IGNITER) {
+            return ACT_DENIED;
+        }
+        break;
+
+    case ARMED_LAUNCH:
+        // every command is available
+        break;
+    }
+
+    bool current_state;
+    err = padstate_get_actstate(state, id, &current_state);
+    if (err) {
+        errno = err;
+        return -1;
+    }
+
+    if (new_state == current_state) {
+        return ACT_OK;
+    }
+
+    err = actuator_set(state, id, new_state);
+    if (err) {
+        errno = err;
+        return -1;
+    }
+    return ACT_OK;
+}
+
+int padstate_get_actstate(padstate_t *state, uint8_t act_id, bool *act_state) {
+    if (act_id < 0 || act_id >= NUM_ACTUATORS) return -1;
+
     int err;
-    err = pthread_rwlock_wrlock(&state->rw_lock);
+    err = pthread_rwlock_rdlock(&state->rw_lock);
     if (err) return err;
 
-    state->actuators[id] = new_state;
+    *act_state = state->actuators[act_id];
```

Comment:
This is exactly how we should be reading state. With an actuator implementation, you can replace the `bool
actuators[NUM_ACTUATORS]` with a `actuator_t actuators[NUM_ACTUATORS]`. Then you can do your
`actuator_controllers_init()` logic inside the `padstate_init` function.

Reply:
Ok got it. I would have to modify actuator_t to have a state field correct?

Reply:
Yes I forgot it currently doesn't have one!

Reply:
Ok, that explains a lot of things